README.md
qimhangul-qt5
=============

## 소개
qimhangul-qt5는 Qt 5를 위한 한글 입력 모듈입니다. Qt 5의 입력기 지원 방식은 Qt
4와 비교하였을 때 다음과 같은 차이가 있습니다:

* xim 지원 중단.
* 입력기 플러그인을 만들 때 필요한 클래스가 QInputContext에서
  QPlatformInputContext로 변경됨. QPlatformInputContext는 QPA (Qt Platform
Abstraction)의 일부로 제공되기 때문에 더 이상 공식적인 클래스로 지원되지 않음.

qimhangul-qt5는 [qimhangul](https://github.com/choehwanjin/qimhangul)의 코드를
기반으로 합니다.

## 빌드 및 설치
빌드하기 위해서는 libhangul 패키지 0.0.4 버전 이상이 필요합니다.
libhangul은 다음 링크에서 구하실 수 있습니다.
http://kldp.net/projects/hangul/

## 빌드와 설치 방법

    $ mkdir build
    $ cd build
    $ qmake ../
    $ make
    # sudo make install


## 설정 방법
Qt 4와 Qt 5가 공존하는 환경이라면 환경 변수 QT4_IM_MODULE 및 QT_IM_MODULE이 각각
Qt 4와 5의 입력기 설정을 담당합니다. 만약 현재 사용 중인 환경에서 환경 변수가
위와 다르다면 알맞게 조정하십시오.

만약 Qt 5의 입력기 환경 변수가 QT_IM_MODULE이라면 아래와 같이 설정할 수
있습니다.

* QT_IM_MODULE=hangul:2 - 두벌식
* QT_IM_MODULE=hangul:3f - 세벌식 최종
* QT_IM_MODULE=hangul:39 - 세벌식 최종

이 외 libhangul에서 사용하는 자판을 사용하려면 hangul: 다음 libhangul 소스의
hangul/hangulinputcontext.c의 HangulKeyboard 구조체의 두 번째 인자를 지정하면
됩니다.

## 사용 방법
한영 전환키는 Shift+Space, Hangul키 입니다.
한자 변환키는 F9, Hangul_Hanja 키 입니다.


NEWS
qimhangul-qt5 0.0.1

 * Qt5 지원
 * qmake 사용
 * libhangul 0.0.12 지원


main.h
/* qimhangul-qt5 - Qt input context plugin for Hangul
 * Based on qimhangul
 * Copyright (C) 2004 Choe Hwanjin
 * Copyright (C) 2015 Shinjo Park
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 */

#include <qpa/qplatforminputcontextplugin_p.h>
#include <QStringList>
#include "qimhangulplatforminputcontext.h"

class QHangulPlatformInputContextPlugin: public QPlatformInputContextPlugin {
    Q_OBJECT
public:
    Q_PLUGIN_METADATA(IID QPlatformInputContextFactoryInterface_iid FILE "qimhangul-qt5.json")

    QHangulPlatformInputContextPlugin();
    ~QHangulPlatformInputContextPlugin();

//    QStringList keys() const;
    QHangulPlatformInputContext *create(const QString &system, const QStringList& paramList);
};


candidatelist.cpp
/* qimhangul - Qt input module for hangul
 * Copyright (C) 2004 Choe Hwanjin
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 */

#include <QBoxLayout>
#include <QFont>
#include <QFrame>
#include <QGridLayout>
#include <QKeyEvent>
#include <QLabel>
#include <QMenu>
#include <QTreeWidget>

#include <hangul.h>

#include "candidatelist.h"

#include <QDebug>

CandidateList::CandidateList() :
    m_list(NULL),
    m_selected(false),
    m_size(0),
    m_currentPage(0),
    m_current(0),
    m_frame(NULL)
{
    m_itemsPerPage = 9;

    m_frame = new QFrame(NULL, Qt::Dialog | Qt::X11BypassWindowManagerHint);
    m_frame->setFrameStyle(QFrame::Panel | QFrame::Plain);

    QBoxLayout *vlayout = new QVBoxLayout(m_frame);
    vlayout->setSpacing(0);
    vlayout->setMargin(3);

    QGridLayout* glayout = new QGridLayout();
    glayout->setSpacing(0);
    glayout->setMargin(0);
    glayout->setColumnStretch(2, 1);
    vlayout->addLayout(glayout);

    m_indexes = new QLabel*[m_itemsPerPage];
    m_values = new QLabel*[m_itemsPerPage];
    m_comments = new QLabel*[m_itemsPerPage];

    for (int i = 0; i < m_itemsPerPage; i++) {
	m_indexes[i] = new QLabel(QString().sprintf(" %d ", i + 1), m_frame);
	m_indexes[i]->setAutoFillBackground(true);
	m_indexes[i]->setMargin(3);
	glayout->addWidget(m_indexes[i], i, 0);

	m_values[i] = new QLabel(m_frame);
	m_values[i]->setAutoFillBackground(true);
	m_values[i]->setMargin(3);
	glayout->addWidget(m_values[i], i, 1);

	m_comments[i] = new QLabel(m_frame);
	m_comments[i]->setWordWrap(true);
	m_comments[i]->setAutoFillBackground(true);
	m_comments[i]->setMargin(3);
	glayout->addWidget(m_comments[i], i, 2);
    }

    QFont font(m_values[0]->font());
    if (font.pointSize() < 0)
	font.setPixelSize(font.pixelSize() * 15 / 10);
    else
	font.setPointSize(font.pointSize() * 15 / 10);

    for (int i = 0; i < m_itemsPerPage; i++) {
	m_values[i]->setFont(font);
    }

    m_statusbar = new QLabel(m_frame);
    m_statusbar->setAlignment(Qt::AlignRight);
    vlayout->addWidget(m_statusbar);
}

CandidateList::~CandidateList()
{
    delete[] m_indexes;
    delete[] m_values;
    delete[] m_comments;

    delete m_frame;
}

void CandidateList::open(const HanjaList* list, int x, int y)
{
    m_list = list;
    m_selected = false;
    m_currentPage = 0;
    m_current = 0;

    if (m_list != NULL) {
	m_size = hanja_list_get_size(m_list);

	updateList();
	updateCursor();

	m_frame->move(x, y);
	m_frame->show();
    }
}

bool CandidateList::isVisible() const
{
    if (m_frame != NULL)
	return m_frame->isVisible();

    return false;
}

QString CandidateList::getCandidate()
{
    return QString::fromUtf8(getCurrent());
}

bool CandidateList::filterEvent(const QKeyEvent *event)
{
    switch (event->key()) {
    case Qt::Key_Up:
    case Qt::Key_K:
    case Qt::Key_Backspace:
	prev();
	break;
    case Qt::Key_Down:
    case Qt::Key_J:
    case Qt::Key_Space:
	next();
	break;
    case Qt::Key_Left:
    case Qt::Key_H:
    case Qt::Key_PageUp:
	prevPage();
	break;
    case Qt::Key_Right:
    case Qt::Key_L:
    case Qt::Key_PageDown:
    case Qt::Key_Tab:
	nextPage();
	break;
    case Qt::Key_Return:
	close();
	m_selected = true;
	return true;
    case Qt::Key_1:
    case Qt::Key_2:
    case Qt::Key_3:
    case Qt::Key_4:
    case Qt::Key_5:
    case Qt::Key_6:
    case Qt::Key_7:
    case Qt::Key_8:
    case Qt::Key_9:
	setCurrent(event->key() - Qt::Key_1);
	close();
	m_selected = true;
	return true;
    case Qt::Key_Escape:
	close();
	return true;
    default:
	break;
    }
    return false;
}

void CandidateList::close()
{
    if (m_frame != NULL)
	m_frame->hide();
}

void CandidateList::move(int x, int y)
{
    if (m_frame != NULL)
	m_frame->move(x, y);
}

void CandidateList::updateList()
{
    for (int i = 0; i < m_itemsPerPage; ++i) {
	if (i + m_currentPage < m_size) {
	    QString text;
	    const char *value = hanja_list_get_nth_value(m_list,
						     i + m_currentPage);
	    const char *comment = hanja_list_get_nth_comment(m_list,
						     i + m_currentPage);

	    m_values[i]->setText(QString::fromUtf8(value));
	    m_comments[i]->setText(QString::fromUtf8(comment));
	} else {
	    m_values[i]->setText("");
	    m_comments[i]->setText("");
	}
    }
}

void CandidateList::updateCursor()
{
    int c = m_current - m_currentPage;

    for (int i = 0; i < m_itemsPerPage; i++) {
	if (i == c) {
	    m_indexes[i]->setForegroundRole(QPalette::HighlightedText);
	    m_indexes[i]->setBackgroundRole(QPalette::Highlight);
	    m_values[i]->setForegroundRole(QPalette::HighlightedText);
	    m_values[i]->setBackgroundRole(QPalette::Highlight);
	    m_comments[i]->setForegroundRole(QPalette::HighlightedText);
	    m_comments[i]->setBackgroundRole(QPalette::Highlight);
	} else {
	    m_indexes[i]->setForegroundRole(QPalette::Foreground);
	    m_indexes[i]->setBackgroundRole(QPalette::Background);
	    m_values[i]->setForegroundRole(QPalette::Foreground);
	    m_values[i]->setBackgroundRole(QPalette::Background);
	    m_comments[i]->setForegroundRole(QPalette::Foreground);
	    m_comments[i]->setBackgroundRole(QPalette::Background);
	}
    }

    QString statusText;
    statusText.sprintf("%d/%d", m_current + 1, m_size);
    m_statusbar->setText(statusText);
}

void
CandidateList::prev()
{
    if (m_current > 0)
	m_current--;

    if (m_current < m_currentPage) {
	m_currentPage -= m_itemsPerPage;
	updateList();
    }
    updateCursor();
}

void
CandidateList::next()
{
    if (m_current < m_size - 1)
	m_current++;

    if (m_current >= m_currentPage + m_itemsPerPage) {
	m_currentPage += m_itemsPerPage;
	updateList();
    }
    updateCursor();
}

void
CandidateList::prevPage()
{
    if (m_currentPage - m_itemsPerPage >= 0) {
	m_current -= m_itemsPerPage;
	if (m_current < 0)
	    m_current = 0;
	m_currentPage -= m_itemsPerPage;
	updateList();
    }
    updateCursor();
}

void
CandidateList::nextPage()
{
    if (m_currentPage + m_itemsPerPage < m_size) {
	m_current += m_itemsPerPage;
	if (m_current > m_size - 1)
	    m_current = m_size - 1;
	m_currentPage += m_itemsPerPage;
	updateList();
    }
    updateCursor();
}

const char*
CandidateList::getCurrent()
{
    return hanja_list_get_nth_value(m_list, m_current);
}

void
CandidateList::setCurrent(int index)
{
    if (m_currentPage + index < m_size) {
	m_current = m_currentPage + index;
    }
}

const char*
CandidateList::getNth(int index)
{
    index += m_currentPage;
    if (index < 0 && index >= m_size)
	return NULL;

    return hanja_list_get_nth_value(m_list, index);
}


qimhangulplatforminputcontext.h
/* qimhangul-qt5 - Qt input context plugin for Hangul
 * Based on qimhangul
 * Copyright (C) 2004 Choe Hwanjin
 * Copyright (C) 2015 Shinjo Park
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 */

#ifndef QIMHANGULPLATFORMINPUTCONTEXT_H
#define QIMHANGULPLATFORMINPUTCONTEXT_H

#include <qpa/qplatforminputcontext.h>
#include <QtCore/QList>
#include <QKeyEvent>

#include <hangul.h>

#include "candidatelist.h"

class QHangulPlatformInputContext: public QPlatformInputContext {
    Q_OBJECT

public:
    QHangulPlatformInputContext();
    QHangulPlatformInputContext(const QString &keyboardType);
    ~QHangulPlatformInputContext();

    bool isValid() const;

    void reset();
    void update(Qt::InputMethodQueries);
    bool filterEvent(const QEvent *event);

    void setFocusObject(QObject *object);

    static HanjaTable *hanjaTable;

private:
    typedef enum {
        MODE_NONE,
        MODE_DIRECT,
        MODE_HANGUL
    } InputMode;

    QString getPreeditString() const;
    QString getCommitString() const;
    void updatePreedit(const QString &str);

    void commitText(const QString &str);
    bool isTriggerKey(const QKeyEvent *event);
    bool isCandidateKey(const QKeyEvent *event);
    bool backspace();
    bool popupCandidateList();
    void setModeInfo(int mode);

    HangulInputContext *m_hic;
    InputMode m_mode;
    QObject *m_focusObject;
    CandidateList *m_candidateList;
};

#endif


qimhangulplatforminputcontext.cpp
/* qimhangul-qt5 - Qt input context plugin for Hangul
 * Based on qimhangul
 * Copyright (C) 2004 Choe Hwanjin
 * Copyright (C) 2015 Shinjo Park
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 */

#include "qimhangulplatforminputcontext.h"
#include <QCoreApplication>
#include <QString>
#include <QTextCodec>
#include <QEvent>
#include <QKeyEvent>
#include <QTextFormat>
#include <QWidget>

#include <QDebug>

#include <hangul.h>

HanjaTable* QHangulPlatformInputContext::hanjaTable = NULL;

static inline QString ucsToQString(const ucschar *ucs);

static bool
onTransition(HangulInputContext* /* hic */,
	     ucschar /* c */, const ucschar *str, void* /* data */)
{
    QTextCodec *codec = QTextCodec::codecForLocale();
    QString s = ucsToQString(str);
    return codec->canEncode(s);
}

static inline QString ucsToQString(const ucschar *ucs) {
    QString str;

    if (ucs != NULL) {
	while (*ucs != 0)
	    str += QChar(*ucs++);
    }
    return str;
}

QHangulPlatformInputContext::QHangulPlatformInputContext() {
    QHangulPlatformInputContext(QStringLiteral("2"));
}

QHangulPlatformInputContext::QHangulPlatformInputContext(const QString &keyboardType) {
    int i, n;
    n = hangul_ic_get_n_keyboards();

    m_hic = NULL;
    for(i = 0; i < n; i++) {
        const char *s = hangul_ic_get_keyboard_id(i);
        if (keyboardType == s) {
            m_hic = hangul_ic_new(s);
            break;
        }
    }

    if (m_hic == NULL) {
        m_hic = hangul_ic_new("2");
    }
    hangul_ic_connect_callback(m_hic, "transition", (void*)onTransition, NULL);
    m_candidateList = NULL;
    m_focusObject = NULL;
}

QHangulPlatformInputContext::~QHangulPlatformInputContext() {
    if (m_hic != NULL) {
        hangul_ic_delete(m_hic);
    }
}

bool QHangulPlatformInputContext::isValid() const {
    return true;
}

void QHangulPlatformInputContext::reset() {
    if (m_candidateList != NULL && m_candidateList->isVisible()) {
	m_candidateList->close();
    }

    const ucschar *flushed = hangul_ic_flush(m_hic);

    // we do not send preedit update IMEvent
    // because commit() send InputMethodEnd event and it remove preedit string

    QString commitString = ucsToQString(flushed);
    if (!commitString.isEmpty()) {
	commitText(commitString);
    } else {
	updatePreedit("");
    }
}

void QHangulPlatformInputContext::update(Qt::InputMethodQueries q) {
    QPlatformInputContext::update(q);
}

QString QHangulPlatformInputContext::getPreeditString() const {
    return ucsToQString(hangul_ic_get_preedit_string(m_hic));
}

QString QHangulPlatformInputContext::getCommitString() const {
    return ucsToQString(hangul_ic_get_commit_string(m_hic));
}

void QHangulPlatformInputContext::updatePreedit(const QString &str) {
    QTextCharFormat fmt;
    fmt.setUnderlineStyle(QTextCharFormat::SingleUnderline);

    QList<QInputMethodEvent::Attribute> attrs;
    attrs << QInputMethodEvent::Attribute(QInputMethodEvent::TextFormat, 0,
		  str.length(), fmt);

    QInputMethodEvent e(str, attrs);
    if (m_focusObject) QCoreApplication::sendEvent(m_focusObject, &e);
}

void QHangulPlatformInputContext::commitText(const QString &str) {
    QInputMethodEvent e;
    e.setCommitString(str);
    if (m_focusObject) QCoreApplication::sendEvent(m_focusObject, &e);
}

bool QHangulPlatformInputContext::isTriggerKey(const QKeyEvent *event) {
    return (event->key() == Qt::Key_Space && (event->modifiers() & Qt::ShiftModifier) == Qt::ShiftModifier) ||
	   (event->key() == Qt::Key_Hangul);
}

bool QHangulPlatformInputContext::isCandidateKey(const QKeyEvent *event) {
    return (event->key() == Qt::Key_Hangul_Hanja) ||
	   (event->key() == Qt::Key_F9);
}

bool QHangulPlatformInputContext::backspace() {
    bool ret = hangul_ic_backspace(m_hic);
    if (ret) {
	QString str = getPreeditString();
	updatePreedit(str);
    }
    return ret;
}

bool QHangulPlatformInputContext::popupCandidateList() {
    const ucschar *text = hangul_ic_get_preedit_string(m_hic);
    if (text != NULL && *text != 0) {
	QString str;
	str += QChar(text[0]);
	HanjaList *list = hanja_table_match_suffix(hanjaTable, str.toUtf8());

	if (m_candidateList == NULL)
	    m_candidateList = new CandidateList();

	QPoint p(0, 0);

	QWidget *focus = qobject_cast<QWidget *>(m_focusObject);
	if (focus != NULL) {
	    QVariant v = focus->inputMethodQuery(Qt::ImMicroFocus);
	    QRect r = v.toRect();
	    p = focus->mapToGlobal(QPoint(r.right(), r.bottom()));
	}

	m_candidateList->open(list, p.x(), p.y());
    }

    return false;
}


bool QHangulPlatformInputContext::filterEvent(const QEvent *event) {
    if (event->type() != QEvent::KeyPress)
	return false;

    const QKeyEvent *keyevent = static_cast<const QKeyEvent*>(event);
    if (m_candidateList != NULL && m_candidateList->isVisible()) {
	if (m_candidateList->filterEvent(keyevent)) {
	    if (m_candidateList->isSelected()) {
		hangul_ic_reset(m_hic);
		QString candidate(m_candidateList->getCandidate());
		commitText(candidate);
	    }
	    m_candidateList->close();
	}
	return true;
    }

    if (keyevent->key() == Qt::Key_Shift)
	return false;

    if (keyevent->key() == Qt::Key_Backspace)
	return backspace();

    if (isTriggerKey(keyevent)) {
	if (m_mode == MODE_DIRECT) {
	    m_mode = MODE_HANGUL;
	} else {
	    reset();
	    m_mode = MODE_DIRECT;
	}
	setModeInfo(m_mode);

	return true;
    }

    if (isCandidateKey(keyevent)) {
	return popupCandidateList();
    }

    if (keyevent->modifiers() & Qt::ControlModifier ||
	keyevent->modifiers() & Qt::AltModifier ||
	keyevent->modifiers() & Qt::MetaModifier) {
	reset();
	return false;
    }

    if (m_mode == MODE_HANGUL) {
	QString text = keyevent->text();
	if (keyevent->modifiers() & Qt::ShiftModifier)
	    text = text.toUpper();
	else
	    text = text.toLower();

	int ascii = 0;
	if (!text.isEmpty())
	    ascii = text[0].unicode();
	bool ret = hangul_ic_process(m_hic, ascii);

	QString commitString = getCommitString();
	if (!commitString.isEmpty())
	    commitText(commitString);

	QString preeditString = getPreeditString();
	if (!preeditString.isEmpty())
	    updatePreedit(preeditString);

	return ret;
    }

    return false;
}

void QHangulPlatformInputContext::setFocusObject(QObject *object) {
    m_focusObject = object;
}


qimhangulplatforminputcontext_x11.cpp
/* qimhangul-qt5 - Qt input context plugin for Hangul
 * Based on qimhangul
 * Copyright (C) 2004 Choe Hwanjin
 * Copyright (C) 2015 Shinjo Park
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 */

#include <ctype.h>

#include <QWidget>
#include <QX11Info>

#include "hangul.h"
#include "qimhangulplatforminputcontext.h"

#include <X11/Xlib.h>

void QHangulPlatformInputContext::setModeInfo(int mode)
{
    Display* display = NULL;
    int screen = 0;

    display = QX11Info::display();
    screen = QX11Info::appScreen();

    if (display != NULL) {
	long data = mode;
	Window window = RootWindow(display, screen);
	Atom status = XInternAtom(display, "_HANGUL_INPUT_MODE", False);
	Atom type = XInternAtom(display, "INTEGER", False);

	if (window != None && status != None && type != None) {
	    XChangeProperty(display, window, status, type,
		    32, PropModeReplace, (const unsigned char *)&data, 1);
	}
    }
}


main.cpp
/* qimhangul-qt5 - Qt input context plugin for Hangul
 * Based on qimhangul
 * Copyright (C) 2004 Choe Hwanjin
 * Copyright (C) 2015 Shinjo Park
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 */

#include "main.h"

QHangulPlatformInputContextPlugin::QHangulPlatformInputContextPlugin() {
    if (QHangulPlatformInputContext::hanjaTable)
	hanja_table_delete(QHangulPlatformInputContext::hanjaTable);
    QHangulPlatformInputContext::hanjaTable = hanja_table_load(NULL);
}

QHangulPlatformInputContextPlugin::~QHangulPlatformInputContextPlugin() {
    if (QHangulPlatformInputContext::hanjaTable)
	hanja_table_delete(QHangulPlatformInputContext::hanjaTable);
}

// If QT_IM_MODULE contains colon, the values after colon will be passed into
// paramList.
// Example: QT_IM_MODULE=hangul:2 -> system=hangul, paramList=[ "2" ]

/*
QStringList QHangulPlatformInputContextPlugin::keys() const {
    return QStringList(QStringLiteral("hangul"));
}
*/

QHangulPlatformInputContext *QHangulPlatformInputContextPlugin::create(const QString &system, const QStringList &paramList) {
    if (system.compare(system, QStringLiteral("hangul"), Qt::CaseInsensitive) == 0) {
        if (paramList.size() == 0) {
            return new QHangulPlatformInputContext();
        } else {
            return new QHangulPlatformInputContext(paramList[0]);
        }
    }
    return 0;
}

